tools/castle-editor/code/castleundosystem.pas

```diff
@@ -0,0 +1,276 @@
+  TUndoData = string;
+  { Reference to the component selected at the moment of recording Undo data
+    Currently we are "finding" the component by name to select it }
+  TSelectedComponent = string;
+
+type
+  { A single change in scene content }
+  TUndoHistoryElement = class(TObject)
+    { Content of this undo }
+    Data: TUndoData;
+    { Compopent, selected at the time of undo recording }
+    Selected: TSelectedComponent;
+    { Index of the value selected/edited in the ObjectInspector }
+    ItemIndex: Integer;
+    { Index of the tab, opened at the moment of the Undo recording }
+    TabIndex: Integer;
+    { Human-readable explanation of what action is recorded in this undo record }
+    Comment: String;
+    { Estimate of this undo record size;
+      It may be several bytes inaccurate,
+      but it's used only to avoid exceeding some preset RAM limit }
+    function Size: SizeInt;
+  end;
+
+type
+  { All undo records currently availalbe }
+  TUndoHistory = specialize TObjectList<TUndoHistoryElement>;
+
+type
+  { Manages undo records and performs undo/redo }
+  TUndoSystem = class(TComponent)
+  //private const MaxUndo = 50;
+  { Max amount of memory that can be reserved by Undo System
+    If the current records total size overruns this limit,
+    the fist undo records will be purged
+    Maybe we'll want to have it as user-definable variable in Castle-Editor settings }
+  private const MaxUndoHistorySize = 128 * 1024 * 1024; //128 Mb
+  private
+    { Current undo step
+      equals to the last element of the Undo History, unless an undo has been performed }
+    CurrentUndo: Integer;
+    { All undo records in this session }
+    UndoHistory: TUndoHistory;
+    { Calculates total size of RAM used by the Undo History }
+    function UndoHistorySize: Integer;
+    { Construct human-readable comment for this Undo record }
+    function GetUndoComment(const UndoI: Integer): String;
+  public
+    { Called when Undo information displayed to the User should be changed }
+    OnUpdateUndo: TNotifyEvent;
+    { Should the Undo be recorded on Release event?
+      Note, that we rely here on Release event, which may never come in case
+      user has switched a window (e.g. with Alt-Tab) while dragging
+      This value will be purged when recording a new Undo record }
+    ScheduleRecordUndoOnRelease: Boolean;
+    constructor Create(AOwner: TComponent); override;
+    destructor Destroy; override;
+    { Should be called after every significant change, including initial loading of the design }
+    { Try to record a new Undo record
+      If several actions have been undone before, all the redo history will be cleared at this moment
+      If the new Undo record is equal to the recorded one then nothing will be recorded }
+    procedure RecordUndo(const UndoData: TUndoData; const SelectedComponent: TSelectedComponent; const ItemIndex: Integer; const TabIndex: Integer; const UndoComment: String);
```

Comment:
Minor comment about English comments: end sentences with a comma, e.g. 

```
    { Try to record a new Undo record.
      If several actions have been undone before, all the redo history will be cleared at this moment.
      If the new Undo record is equal to the recorded one then nothing will be recorded. }
```

This goes also for 1-sentence descriptions, like

```
    { If it's possible to perform Undo right now. }
    function IsUndoPossible: Boolean;
```

( Also "If it's possible to Xxx" -> "Is it possible to Xxx" I think sounds a bit better. )

Reply:
Should be fixed now.